go/pkg/bertyprotocol/store_message.go

```diff
@@ -40,7 +44,80 @@ func (m *messageStore) setLogger(l *zap.Logger) {
 	m.logger = l.With(zap.String("group-id", fmt.Sprintf("%.6s", base64.StdEncoding.EncodeToString(m.g.PublicKey))))
 }
 
-func (m *messageStore) openMessage(ctx context.Context, e ipfslog.Entry) (*bertytypes.GroupMessageEvent, error) {
+func (m *messageStore) openCachedMessages(devicePK []byte) bool {
+	select {
+	case m.cacheChan <- devicePK:
+		return true
+	default:
+		return false
+	}
+}
+
+// Add the event into a slice of the map
+func (m *messageStore) addToCache(devicePK []byte, e ipfslog.Entry) {
+	m.logger.Debug("addToCache", zap.Any("devicePK", devicePK), zap.Any("event", e))
+	m.cacheLock.Lock()
+	m.cache[string(devicePK)] = append(m.cache[string(devicePK)], e)
+	m.cacheLock.Unlock()
+
+	if !m.openCachedMessages(devicePK) {
+		m.logger.Error("Channel is full, can't queued the devicePK for scanning")
+	}
+}
+
+// Delete the event from the map, or if there is no more event in the slice, delete the map entry
+// The event order is not important, so we swap events for deleting
+func (m *messageStore) unsafeDeleteFromCache(devicePK []byte, e ipfslog.Entry) {
+	events := m.cache[string(devicePK)]
+	for i, event := range events {
+		if event == e {
+			events[i] = events[len(events)-1]
+			break
+		}
+	}
+	if len(events) == 0 {
+		delete(m.cache, string(devicePK))
+	} else {
+		m.cache[string(devicePK)] = events[:len(events)-1]
+	}
+}
+
+// openFromCache tries to open messages for a given devicePK in a regular interval of time, until there is no message no more
+func (m *messageStore) openFromCache(ctx context.Context, devicePK []byte) {
+	m.cacheLock.Lock()
+	if events, ok := m.cache[string(devicePK)]; ok {
+		for _, e := range events {
+			m.logger.Debug("openFromCache: processing event", zap.Any("devicePK", devicePK), zap.Any("event", e))
+			evt, err := m.openMessage(ctx, e, false)
+			if err != nil {
+				m.logger.Error("unable to open message", zap.Error(err))
+				continue
+			}
+
+			m.Emit(ctx, evt)
+			m.unsafeDeleteFromCache(devicePK, e)
+		}
+		if len(events) == 0 {
```

Comment:
this block of if statement is redudant?

Reply:
oh yes, there was a for loop here, and I forget to remove this block !